bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/TraversePageHandler.java

```diff
@@ -28,30 +28,84 @@
  * @since 3.5
  */
 public class TraversePageHandler extends WidgetMethodHandler {
-
 	/**
 	 * The parameters for traverse(int).
 	 */
 	private static final Class<?>[] METHOD_PARAMETERS = { int.class };
-
 	@Override
 	public final Object execute(final ExecutionEvent event) {
 		Control focusControl = Display.getCurrent().getFocusControl();
 		if (focusControl != null) {
-			int traversal = "next".equals(methodName) ? SWT.TRAVERSE_PAGE_NEXT : SWT.TRAVERSE_PAGE_PREVIOUS; //$NON-NLS-1$
+			boolean forward = "next".equals(methodName); //$NON-NLS-1$
+			int traversal = getTraversalDirection(forward);
 			Control control = focusControl;
+
 			do {
+				if (control instanceof CTabFolder folder && isFinalItemInCTabFolder(folder, forward)
+						&& !areHiddenItems(folder)) {
+					loopToSecondToFirstItemInCTabFolder(folder, forward);
+					traversal = getTraversalDirection(!forward); // we are in the second-to-last item in the given
+					// direction. Now, use the Traverse-event to move back by one
+				}
 				if (control.traverse(traversal))
 					return null;
 				if (control instanceof Shell)
 					return null;
 				control = control.getParent();
 			} while (control != null);
 		}
-
 		return null;
 	}
 
+	private boolean areHiddenItems(CTabFolder folder) {
+		CTabItem[] items = folder.getItems();
+		for (CTabItem i : items) {
```

Comment:
use this https://stackoverflow.com/questions/31419931/how-do-you-check-if-a-list-contains-an-element-that-matches-some-predicate

Reply:
done